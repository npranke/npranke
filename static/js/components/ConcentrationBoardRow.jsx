import PropTypes from 'prop-types'
import React from 'react'

export function ConcentrationBoardRow(props) {
    const rowPictures = props.pictures.map((picture) => {
        const pictureName = Object.keys(picture)[0]

        const { id, pictureid } = picture[pictureName]

        let pictureImage
        if (props.first.id === id || props.second.id === id) {
            pictureImage = picture[pictureName].front
        } else if (props.matches.indexOf(pictureid) !== -1) {
            pictureImage = picture[pictureName].matched
        } else {
            pictureImage = picture[pictureName].back
        }

        return (
            <React.Fragment key={ id }>
                { pictureImage }
            </React.Fragment>
        )
    })

    return (
        <div className="board-row">
            { rowPictures }
        </div>
    )
}

ConcentrationBoardRow.defaultProps = {
    first: {
        id: null,
        pictureid: null,
    },
    matches: [],
    second: {
        id: null,
        pictureid: null,
    },
}

ConcentrationBoardRow.propTypes = {
    first: PropTypes.shape({
        id: PropTypes.string,
        pictureid: PropTypes.string,
    }),
    matches: PropTypes.arrayOf(PropTypes.string),
    pictures: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.string,
            pictureid: PropTypes.string,
            back: PropTypes.node,
            front: PropTypes.node,
            matched: PropTypes.node,
        }),
    ).isRequired,
    second: PropTypes.shape({
        id: PropTypes.string,
        pictureid: PropTypes.string,
    }),
}

export default ConcentrationBoardRow
