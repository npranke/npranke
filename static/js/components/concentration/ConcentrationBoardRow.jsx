import PropTypes from 'prop-types'
import React from 'react'

function ConcentrationBoardRow(props) {
    const rowPictures = props.pictures.map((picture) => {
        const pictureName = Object.keys(picture)[0]

        const { id, pictureid } = picture[pictureName]

        let pictureImage
        if (props.first.id === id || props.second.id === id) {
            pictureImage = picture[pictureName].front
        } else if (props.matches.indexOf(pictureid) !== -1) {
            pictureImage = picture[pictureName].matched
        } else {
            pictureImage = picture[pictureName].back
        }

        return (
            <React.Fragment key={ id }>
                { pictureImage }
            </React.Fragment>
        )
    })

    return (
        <div className="board-row">
            { rowPictures }
        </div>
    )
}

ConcentrationBoardRow.defaultProps = {
    first: {
        id: null,
        pictureid: null,
    },
    matches: [],
    second: {
        id: null,
        pictureid: null,
    },
}

ConcentrationBoardRow.propTypes = {
    first: PropTypes.shape({
        id: PropTypes.string,
        pictureid: PropTypes.string,
    }),
    matches: PropTypes.arrayOf(PropTypes.string),
    pictures: PropTypes.arrayOf(
        PropTypes.objectOf(
            PropTypes.shape({
                id: PropTypes.string.isRequired,
                pictureid: PropTypes.string.isRequired,
                back: PropTypes.node.isRequired,
                front: PropTypes.node.isRequired,
                matched: PropTypes.node.isRequired,
            }).isRequired,
        ).isRequired,
    ).isRequired,
    second: PropTypes.shape({
        id: PropTypes.string,
        pictureid: PropTypes.string,
    }),
}

export default ConcentrationBoardRow
