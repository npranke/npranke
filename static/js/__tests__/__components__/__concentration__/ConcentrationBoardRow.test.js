import Adapter from 'enzyme-adapter-react-16'
import Enzyme, { shallow } from 'enzyme'
import React from 'react'

import ConcentrationBoardRow from
    '@components/concentration/ConcentrationBoardRow'

Enzyme.configure({ adapter: new Adapter() })

const pictures = Array.from({ length: 6 }, (value, index) => {
    return `${index}`
}).map((pictureid) => {
    return {
        [`picture${pictureid}-a`]: {
            id: `${pictureid}-a`,
            pictureid,
            back: <img src="" alt="" className="picture-back" />,
            front: <img src="" alt="" className="picture-front" />,
            matched: <img src="" alt="" className="picture-matched" />,
        },
    }
})

describe('ConcentrationBoardRow', () => {
    test('board-row has pictures', () => {
        const concentrationBoardRow = shallow(
            <ConcentrationBoardRow pictures={ pictures } />,
        )

        expect(
            concentrationBoardRow.find('.board-row').children(),
        ).toHaveLength(6)
    })

    test('has all picture-back when no first, second, matches', () => {
        const concentrationBoardRow = shallow(
            <ConcentrationBoardRow pictures={ pictures } />,
        )

        expect(
            concentrationBoardRow.find('.picture-back'),
        ).toHaveLength(pictures.length)
    })

    test('has one picture-front when matching first', () => {
        const concentrationBoardRow = shallow(
            <ConcentrationBoardRow
                pictures={ pictures }
                first={ { id: '2-a', pictureid: '2' } }
            />,
        )

        expect(
            concentrationBoardRow.find('.picture-front'),
        ).toHaveLength(1)
    })

    test('has no picture-front when non-matching first', () => {
        const concentrationBoardRow = shallow(
            <ConcentrationBoardRow
                pictures={ pictures }
                first={ { id: '9-a', pictureid: '9' } }
            />,
        )

        expect(
            concentrationBoardRow.find('.picture-front'),
        ).toHaveLength(0)
    })

    test('has one picture-front when matching second', () => {
        const concentrationBoardRow = shallow(
            <ConcentrationBoardRow
                pictures={ pictures }
                second={ { id: '3-a', pictureid: '3' } }
            />,
        )

        expect(
            concentrationBoardRow.find('.picture-front'),
        ).toHaveLength(1)
    })

    test('has no picture-front when non-matching second', () => {
        const concentrationBoardRow = shallow(
            <ConcentrationBoardRow
                pictures={ pictures }
                second={ { id: '8-b', pictureid: '8' } }
            />,
        )

        expect(
            concentrationBoardRow.find('.picture-front'),
        ).toHaveLength(0)
    })

    test('has two picture-front when matching first and second', () => {
        const concentrationBoardRow = shallow(
            <ConcentrationBoardRow
                pictures={ pictures }
                first={ { id: '0-a', pictureid: '0' } }
                second={ { id: '3-a', pictureid: '3' } }
            />,
        )

        expect(
            concentrationBoardRow.find('.picture-front'),
        ).toHaveLength(2)
    })

    test('has no picture-front when non-matching first and second', () => {
        const concentrationBoardRow = shallow(
            <ConcentrationBoardRow
                pictures={ pictures }
                first={ { id: '10-b', pictureid: '10' } }
                second={ { id: '7-a', pictureid: '7' } }
            />,
        )

        expect(
            concentrationBoardRow.find('.picture-front'),
        ).toHaveLength(0)
    })

    test('has no picture-matched with no matches', () => {
        const concentrationBoardRow = shallow(
            <ConcentrationBoardRow pictures={ pictures } />,
        )

        expect(
            concentrationBoardRow.find('.picture-matched'),
        ).toHaveLength(0)
    })

    test('has some picture-matched with some matches', () => {
        const concentrationBoardRow = shallow(
            <ConcentrationBoardRow
                pictures={ pictures }
                matches={ ['0', '1', '5'] }
            />,
        )

        expect(
            concentrationBoardRow.find('.picture-matched'),
        ).toHaveLength(3)
    })

    test('has all picture-matched with all matches', () => {
        const concentrationBoardRow = shallow(
            <ConcentrationBoardRow
                pictures={ pictures }
                matches={
                    Array.from(
                        { length: 6 },
                        (value, index) => { return `${index}` },
                    )
                }
            />,
        )

        expect(
            concentrationBoardRow.find('.picture-matched'),
        ).toHaveLength(6)
    })

    test('can have picture-back, picture-front, picture-matched mix', () => {
        const concentrationBoardRow = shallow(
            <ConcentrationBoardRow
                pictures={ pictures }
                matches={ ['0', '5'] }
                first={ { id: '4-a', pictureid: '4' } }
                second={ { id: '2-a', pictureid: '2' } }
            />,
        )

        expect(
            concentrationBoardRow.find('.picture-back'),
        ).toHaveLength(2)
        expect(
            concentrationBoardRow.find('.picture-front'),
        ).toHaveLength(2)
        expect(
            concentrationBoardRow.find('.picture-matched'),
        ).toHaveLength(2)
    })
})

describe('ConcentrationBoardRow snapshot', () => {
    test('matches snapshot', () => {
        const concentrationBoardRow = shallow(
            <ConcentrationBoardRow pictures={ pictures } />,
        )

        expect(concentrationBoardRow).toMatchSnapshot()
    })
})
