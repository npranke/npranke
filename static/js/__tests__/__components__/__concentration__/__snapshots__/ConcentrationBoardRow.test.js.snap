// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConcentrationBoardRow snapshot matches snapshot with all matches 1`] = `
<DocumentFragment>
  <div
    class="board-row"
  >
    <img
      alt=""
      class="picture-matched"
      src=""
    />
    <img
      alt=""
      class="picture-matched"
      src=""
    />
    <img
      alt=""
      class="picture-matched"
      src=""
    />
    <img
      alt=""
      class="picture-matched"
      src=""
    />
    <img
      alt=""
      class="picture-matched"
      src=""
    />
    <img
      alt=""
      class="picture-matched"
      src=""
    />
  </div>
</DocumentFragment>
`;

exports[`ConcentrationBoardRow snapshot matches snapshot with no matches 1`] = `
<DocumentFragment>
  <div
    class="board-row"
  >
    <img
      alt=""
      class="picture-back"
      src=""
    />
    <img
      alt=""
      class="picture-back"
      src=""
    />
    <img
      alt=""
      class="picture-back"
      src=""
    />
    <img
      alt=""
      class="picture-back"
      src=""
    />
    <img
      alt=""
      class="picture-back"
      src=""
    />
    <img
      alt=""
      class="picture-back"
      src=""
    />
  </div>
</DocumentFragment>
`;

exports[`ConcentrationBoardRow snapshot matches snapshot with some matches 1`] = `
<DocumentFragment>
  <div
    class="board-row"
  >
    <img
      alt=""
      class="picture-back"
      src=""
    />
    <img
      alt=""
      class="picture-back"
      src=""
    />
    <img
      alt=""
      class="picture-matched"
      src=""
    />
    <img
      alt=""
      class="picture-back"
      src=""
    />
    <img
      alt=""
      class="picture-matched"
      src=""
    />
    <img
      alt=""
      class="picture-back"
      src=""
    />
  </div>
</DocumentFragment>
`;

exports[`ConcentrationBoardRow snapshot matches snapshot with two picture-front 1`] = `
<DocumentFragment>
  <div
    class="board-row"
  >
    <img
      alt=""
      class="picture-back"
      src=""
    />
    <img
      alt=""
      class="picture-front"
      src=""
    />
    <img
      alt=""
      class="picture-back"
      src=""
    />
    <img
      alt=""
      class="picture-back"
      src=""
    />
    <img
      alt=""
      class="picture-front"
      src=""
    />
    <img
      alt=""
      class="picture-back"
      src=""
    />
  </div>
</DocumentFragment>
`;
